src/OrleansSQLUtils/CreateOrleansTables_PostgreSql.sql

```diff
@@ -859,3 +859,213 @@ VALUES
+	--
+	-- If the version number explicitly returned is still the same, Orleans interprets it so the update did not succeed
+	-- and throws an InconsistentStateException.
+	--
+	-- See further information at http://dotnet.github.io/orleans/Getting-Started-With-Orleans/Grain-Persistence.
+	IF _GrainStateVersion IS NOT NULL
+	THEN
+		UPDATE Storage
+		SET
+			PayloadBinary = _PayloadBinary,
+			PayloadJson = _PayloadJson,
+			PayloadXml = _PayloadXml,
+			ModifiedOn = (now() at time zone 'utc'),
+            Version = Version + 1
+       
+		WHERE
+			GrainIdHash = _GrainIdHash AND _GrainIdHash IS NOT NULL
+			AND GrainTypeHash = _GrainTypeHash AND _GrainTypeHash IS NOT NULL
+			AND GrainIdN0 = _GrainIdN0 AND _GrainIdN0 IS NOT NULL
+			AND GrainIdN1 = _GrainIdN1 AND _GrainIdN1 IS NOT NULL
+			AND GrainTypeString = _GrainTypeString AND _GrainTypeString IS NOT NULL
+			AND ((_GrainIdExtensionString IS NOT NULL AND GrainIdExtensionString IS NOT NULL AND GrainIdExtensionString = _GrainIdExtensionString) OR _GrainIdExtensionString IS NULL AND GrainIdExtensionString IS NULL)
+			AND ServiceId = _ServiceId AND _ServiceId IS NOT NULL
+			AND Version IS NOT NULL AND Version = _GrainStateVersion AND _GrainStateVersion IS NOT NULL;
+            
+		GET DIAGNOSTICS RowCountVar = ROW_COUNT;
+        IF RowCountVar > 0
+        THEN
+			_newGrainStateVersion := _GrainStateVersion + 1;
+        END IF;            
+            
+            
+	END IF;
+
+	-- The grain state has not been read. The following locks rather pessimistically
+	-- to ensure only on INSERT succeeds.
+    IF _GrainStateVersion IS NULL
+	THEN
+		INSERT INTO Storage
+		(
+			GrainIdHash,
+			GrainIdN0,
+			GrainIdN1,
+			GrainTypeHash,
+			GrainTypeString,
+			GrainIdExtensionString,
+			ServiceId,
+			PayloadBinary,
+			PayloadJson,
+			PayloadXml,
+			ModifiedOn,
+			Version
+		)
+		SELECT  
+			_GrainIdHash,
+			_GrainIdN0,
+			_GrainIdN1,
+			_GrainTypeHash,
+			_GrainTypeString,
+			_GrainIdExtensionString,
+			_ServiceId,
+			_PayloadBinary,
+			_PayloadJson,
+			_PayloadXml,
+           (now() at time zone 'utc'),
+			1
+		WHERE NOT EXISTS
+		 (
+			-- There should not be any version of this grain state.
+			SELECT 1
+			FROM Storage
+			WHERE
+				GrainIdHash = _GrainIdHash AND _GrainIdHash IS NOT NULL
+				AND GrainTypeHash = _GrainTypeHash AND _GrainTypeHash IS NOT NULL
+				AND GrainIdN0 = _GrainIdN0 AND _GrainIdN0 IS NOT NULL
+				AND GrainIdN1 = _GrainIdN1 AND _GrainIdN1 IS NOT NULL
+				AND GrainTypeString = _GrainTypeString AND _GrainTypeString IS NOT NULL
+				AND ((_GrainIdExtensionString IS NOT NULL AND GrainIdExtensionString IS NOT NULL AND GrainIdExtensionString = _GrainIdExtensionString) OR _GrainIdExtensionString IS NULL AND GrainIdExtensionString IS NULL)
+				AND ServiceId = _ServiceId AND _ServiceId IS NOT NULL
+		 ); 
+
+		GET DIAGNOSTICS RowCountVar = ROW_COUNT;
+        IF RowCountVar > 0
+        THEN
+			_newGrainStateVersion := 1;
```

Comment:
This is a nit, but the identations seem to be off here. This is unfortunately something I've introduced myself too to other scripts, but if you feel like labouring over this, maybe something to notice.